src/Nip17/DirectMessage.php

```diff
@@ -0,0 +1,275 @@
+     * @param string $message The message content
+     * @param array $additionalTags Additional tags to include
+     * @param string|null $replyToId ID of message being replied to (optional)
+     *
+     * @return array An array containing the created seal, gift wrap, and relay information
+     */
+    public function sendDirectMessage(
+        string $senderPrivkey,
+        string $receiverPubkey,
+        string $message,
+        array $additionalTags = [],
+        ?string $replyToId = null,
+    ): array {
+        // Derive sender's public key from private key
+        $senderPubkey = $this->keyService->getPublicKey($senderPrivkey);
+
+        // Discover receiver's preferred relays
+        $receiverRelays = $this->getPreferredRelaysForPubkey($receiverPubkey);
+
+        // Get sender's preferred relays
+        $senderRelays = $this->getPreferredRelaysForPubkey($senderPubkey);
+
+        // Create the base event (kind 14)
+        $event = $this->createDirectMessageEvent($message, $receiverPubkey, $replyToId, $additionalTags);
+        $event->setSenderPubkey($senderPubkey);
+
+        // // Step 1: Create a seal (kind 13) for the event
+        $sealEvent = $this->giftWrapService->createSeal($event, $senderPrivkey, $receiverPubkey);
+
+        // // Step 2: Create a gift wrap (kind 1059) to further protect the seal
+        $receiverGiftWrap = $this->giftWrapService->createGiftWrap($sealEvent, $receiverPubkey);
+
+        // Also create a copy for the sender to keep track of sent messages
+        $senderGiftWrap = $this->giftWrapService->createGiftWrap($sealEvent, $senderPubkey);
+
+        return [
+            'seal' => $sealEvent,
+            'receiver' => $receiverGiftWrap,
+            'sender' => $senderGiftWrap,
+            'receiver_relays' => $receiverRelays,
+            'sender_relays' => $senderRelays,
+        ];
+    }
+
+    /**
+     * Decrypt a direct message from a gift wrap and seal.
+     *
+     * @param GiftWrapInterface|EventInterface|\stdClass $giftWrap The gift wrapped event to decrypt
+     * @param string $receiverPrivkey The private key of the recipient (you)
+     * @param bool $verifyRecipient Whether to verify that the gift wrap is addressed to the receiver
+     *
+     * @return array|null The decrypted event as an array, or null if decryption failed or recipient verification failed
+     */
+    public static function decryptDirectMessage(
+        $giftWrap,
+        string $receiverPrivkey,
+        bool $verifyRecipient = true,
+    ): ?array {
+        // Extract the recipient pubkey from p tag
+        $receiverPubkey = (new Key())->getPublicKey($receiverPrivkey);
+        $isAddressedToReceiver = false;
+
+        // Check that the gift wrap is addressed to the receiver via p tag
+        // Handle both Event objects and stdClass objects from relays
+        $tags = method_exists($giftWrap, 'getTags') ? $giftWrap->getTags() : ($giftWrap->tags ?? []);
```

Comment:
In which case the `$giftWrap` variable is a object of the GiftWrap class? Why is this `method_exists` check needed? The tags property is protected is should be only accessed via the `getTags()` method of the GiftWrap class.

Reply:
It's in the comment above, in the case the event is a stdClass, a received event is not automatically deserialized as a GiftWrap class. In the nip-17 example where the created gift wrap is a giftwrap it already is an instance of GiftWrap

Reply:
I would make it so that if the object is a stdClass convert it to a Giftwrap (Event) object. Does that make sense?

Reply:
Yes, I think in general it would be good to have helper methods (factory class?) that convert stdClass objects to the correct kind-specific classes.